### PR TITLE
Issue #388: Refactor preprocessing functionality

### LINKED ISSUE #388: Simplifications and refactor of `add_event_vars`
In PR #375 we added a function `add_event_vars`. Now we have this foothold, here is my short write up of how we can refactor it and move to a broader plan about how to handle preprocessing:

My proposal / understanding of what is required:

1. A class structure for data formats
2. Data format to model class converters
3. Helper functions to add columns

### 1. A class structure for data formats

The only one we know of currently is `epidist_linelist`.
The properties of an `epidist_linelist` object are:

1. `xtime_lwr`, `xtime_upr` and `xwindow` columns for `x` is `p` or `s`
2. Does it have time columns? I am unsure and I read others as being unsure
3. Does it have delay columns? Likewise: unsure and I read others as being unsure

```r
as_epidist_linelist <- function() {
  # could call add_ internally
  # could expect user to have called them
  # converts all names
  # uses name = NULL as argument pattern that we have currently
}
```

A basic attempt to start this just changing the names.
I think we should also have checkers in here:

<details>

```r
as_epidist_linelist <- function(
  data, ptime_lwr = NULL, pwindow = NULL, ptime_upr = NULL, stime_lwr = NULL,
  swindow = NULL, stime_upr = NULL
) {
  class(data) <- c("epidist_linelist", class(data))
  data <- .rename_column(data, "ptime_lwr", ptime_lwr)
  data <- .rename_column(data, "ptime_upr", ptime_upr)
  data <- .rename_column(data, "stime_lwr", stime_lwr)
  data <- .rename_column(data, "stime_upr", stime_upr)
  data <- .rename_column(data, "pwindow", pwindow)
  data <- .rename_column(data, "swindow", swindow)
  return(data)
}
```

Note: how to do `.rename_column` in a way which doesn't involve copying many times?

</details>

The other formats are:

* Some kind of format for aggregated counts `epidist_counts`
* This could be it?

At meta level I think for time being we can put things in either format or model structure. Seeing as we only have one format and one model. Over time if it becomes clear things should really go in one or the other we can move things around.

Can we actually go from different formats into the same model? Probably not with `latent_individual` as it's quite "maximal information". Maybe we can go from `linelist` to models designed for more count style data as it's a compression of information. Why would you do this? Not sure. It's cheap / quick

### 2. Data format to model class converters

This is the only example we have at the moment:

```r
as_latent_indvidual.epidist_linelist <- function() {
  # convert to latent_individual from epidist_linelist
  # which things are latent_individual specific?
  # could call add_ functions here
}
```

```r
as_model.epidist_format <- function() {
 # this is a template for other format to model convertors
}
```

### 3. Helper functions to add columns

We either expect 1. users to call these functions, 2. us to call these functions inside the above, 3. a mixture... [Adam to research or prompt others to find out].

```r
add_event_vars <- function() {
  # helper, do one thing, don't care about names, add columns
  # therefore, user must provide names of columns to manipulate
  # could either add columns with set name or let user set the name
  # seems nonsense if let user set name then convert straight away
  # having output of this be df with mix of user names and our names seems bad
}
```

An attempt to flesh this out looks as follows:

```r
add_event_vars <- function(
  data, ptime_lwr = NULL, pwindow = NULL, stime_lwr = NULL, swindow = NULL
) {
  data |>
    mutate(
      ptime_upr = !!ptime_lwr + !!pwindow,
      stime_upr = !!stime_lwr + !!swindow
    )
}

```

The other two helpers I think we need are:

```r
add_obs_vars <- function() {
  # ...
}

add_delay_vars <- function() {
  # ...
}
```

An attempt to flesh out the first highlights some challenges:

```r
add_obs_vars <- function(
 data, obs_time, ptime_lwr = NULL
) {
  # obs_time could be numeric (same for all cases) or vector (different)
  # do we need to give the name for ptime_lwr here? no guaruntee of what it is
  data |>
    mutate(
      obs_time = obs_time,
      relative_obs_time = obs_time - ptime_lwr
    )
}
```

### PR DESCRIPTION
**Pull request by athowes**
## Description

> [!IMPORTANT]
> **Current review ask** ready for review

This PR will close #388.

Summary of what has been done:

* Add `epidist_linelist` class
* Add function `as_epidist_linelist` as date based entry point to `epidist_linelist` class
* Added ad-hoc function `as_epidist_linelist_time` as temporary time based entry point to `epidist_linelist` class
* Changed validation approach to use `epidist_validate_data` for data classes (like `epidist_linelist`) and `epidist_validate_model` for model classes (like `latent_individual`)

Summary of remains to do:

* Update vignettes to run
  * [x] Ebola vignette
  * [x] Approximate inference vignette
  * [x] Get started vignette
  * [x] FAQ
* Update text in vignettes
  * [x] Ebola vignette
  * [x] Approximate inference vignette
  * [x] Get started vignette
  * [x] FAQ

Summary of new issues / extensions:

1. https://github.com/epinowcast/epidist/issues/415 `as_epidist_linelist` could accept a user not passing an upper bound (and could assume that no upper bound means daily censoring)
1. https://github.com/epinowcast/epidist/issues/412 Think about other classes like `epidist_linelist` 
1. https://github.com/epinowcast/epidist/issues/414 Provide more entry pathways to `epidist_linelist` including a better thought out time entry point 
1. https://github.com/epinowcast/epidist/issues/413 Right now `epidist_linelist` doesn't include a column for the "case" -- should we enforce that users provide this? It could be useful to have to produce plots and things
1. https://github.com/epinowcast/epidist/issues/416 RIght now `as_latent_individual` does some making new columns. Can we refactor these into helpers as was original plan?
1. #402  In the class constructor for `direct_model` they are called `ptime` and `stime` and use a `data.frame` without going via `epidist_linelist` -- need to sort this out to go via `direct_model`

## Checklist

- [x] My PR is based on a package issue and I have explicitly linked it.
- [x] I have included the target issue or issues in the PR title in the for Issue(s) *issue-numbers*: PR title
- [x] I have read the [contribution guidelines](https://github.com/epinowcast/.github/blob/main/CONTRIBUTING.md).
- [x] I have tested my changes locally.
- [x] I have added or updated unit tests where necessary.
- [x] I have updated the documentation if required.
- [x] My code follows the established coding standards.
- [ ] I have added a news item linked to this PR.
- [x] I have reviewed CI checks for this PR and addressed them as far as I am able.

### CODE REVIEW

File: R/preprocess.R [hunk 1]
```diff
@@ -1,69 +1,36 @@
...
+      ptime_upr = !!ptime_lwr + !!pwindow,
+      stime_upr = !!stime_lwr + !!swindow
+    )
+}
 
-  linelist <- dplyr::relocate(
-    linelist, ptime_lwr, ptime_upr, pwindow, stime_lwr, stime_upr, swindow
-  )
+as_epidist_linelist <- function(
+  data, ptime_lwr = NULL, pwindow = NULL, ptime_upr = NULL, stime_lwr = NULL,
```
Comment (seabbs) [inline R/preprocess.R:12]: unclear why you would need pwindow and swindow as inputs vs creating here. 

I thought we discussed making the input dates

Comment (athowes) [reply R/preprocess.R:12]: > unclear why you would need pwindow and swindow as inputs vs creating here.

We discussed that `as_epidist_linelist` would only do renaming. Remains to determine how users input `pwindow` and `swindow`

> I thought we discussed making the input dates

Yes we did. I thought we discussed `as_epidist_linelist` only doing renaming and other helper functions doing creation of columns.

File: R/preprocess.R [hunk 2]
```diff
@@ -1,69 +1,45 @@
...
-  )
+add_delay_vars <- function() {
+  return(NULL)
+}
 
-  linelist <- dplyr::relocate(
-    linelist, ptime_lwr, ptime_upr, pwindow, stime_lwr, stime_upr, swindow
-  )
+as_epidist_linelist <- function(
+    data, ptime_lwr = NULL, pwindow = NULL, ptime_upr = NULL, stime_lwr = NULL,
```
[the remark this answers is not in the document]
Comment (athowes) [reply R/preprocess.R:28]: Could have date as another column to allow reconstruction

[the remark this answers is not in the document]
Comment (athowes) [reply R/preprocess.R:28]: 1. Don't allow anything other than date
2. One day different if it's a one day window or the same day
3. Still require `pdate_lwr` and `pdate_upr` even if they are the same date
4. Could later extend to accepting `NULL` as `pdate_upr`
5. Check `datetime`
6. Should `as_epidist_linelist` include the observation times
  * Must include events
  * Could include delay (direct)
  * Could include times (direct and latent individual)
7. Observation date as input
  * Yes do this

[the remark this answers is not in the document]
Comment (athowes) [reply R/preprocess.R:28]: Look for `case` and `row_id` in the repo and if not used then get rid. Make new issue for handling of `case`. Needed in some plots

**Comment by athowes** [comment]
> In the class constructor for direct_model I call the times ptime and stime -- this seems debateable

**Comment by seabbs** [comment]
🥳

**Comment by athowes** [comment]
Note that actually `row_id` is used in the `latent_indvidual` model:

```r
    brms::stanvar(
      block = "data",
      scode = "array[N - wN] int noverlap;",
      x = filter(data, woverlap == 0)$row_id,
      name = "noverlap"
    ) +
 ```

**Comment by seabbs** [comment]
As it's a transformed variable I think we can create it in latent individual. As it's an internal variable we might want to use .row_id to avoid closes with user variables

**Comment by athowes** [comment]
Finding a pain point here is changing all the tests to use a date version of simulated data (knowing that probably we will need to revert this as we add functionality to have non-date data).

**Comment by seabbs** [comment]
Sure but just as a helper function in test that?

**Comment by athowes** [comment]
> Sure but just as a helper function in test that?

It's currently a regular packaged.

It'll need to be used in some of the vignettes as well so would prefer to keep it as a packaged function.

We could also consider putting it as an internal function and using `epidist:::` in the vignettes.

**Comment by athowes** [comment]
> Sure but just as a helper function in test that?

Note that I think we are going to need this functionality eventually anyway, so if this is something we want to think about and implement here I could be open to it.

**Comment by athowes** [comment]
In that case I'm in favour of adding it as an exported function (as is currently) so that it can be used in the vignettes.

Then that would give us chance to:

> [implement] here quickly and think about the design in another issue as I don't think we want to rush it.

File: R/preprocess.R [hunk 3]
```diff
@@ -1,69 +1,88 @@
...
-  linelist <- .rename_column(linelist, "stime_upr", stime_upr)
-  linelist <- .rename_column(linelist, "pwindow", pwindow)
-  linelist <- .rename_column(linelist, "swindow", swindow)
+  class(data) <- c("epidist_linelist", class(data))
 
-  if (is.numeric(pwindow)) {
-    cli::cli_warn("Overwriting using numeric value(s) of pwindow provided!")
-    linelist$pwindow <- pwindow
-  }
+  # Rename columns to our internal names: inefficient and needs refactor
```
Comment (seabbs) [inline R/preprocess.R:17]: This should be an issue not a comment

Comment (athowes) [reply R/preprocess.R:17]: Agree

Comment (athowes) [reply R/preprocess.R]: https://github.com/epinowcast/epidist/issues/417

File: NAMESPACE [hunk 4]
```diff
@@ -17,12 +17,15 @@ S3method(epidist_formula_model,epidist_latent_individual)
...
-export(add_event_vars)
+S3method(epidist_validate_data,default)
+S3method(epidist_validate_data,epidist_linelist)
+S3method(epidist_validate_model,default)
+S3method(epidist_validate_model,epidist_direct_model)
+S3method(epidist_validate_model,epidist_latent_individual)
 export(add_mean_sd)
 export(as_direct_model)
+export(as_epidist_linelist)
+export(as_epidist_linelist_time)
```
[the remark this answers is not in the document]
Comment (athowes) [reply NAMESPACE]: [9ef66f6](https://github.com/epinowcast/epidist/pull/390/commits/9ef66f662da2de643e80bd7b28843d92eaae8a38)

[the remark this answers is not in the document]
Comment (athowes) [reply NAMESPACE]: And [e2e68d2](https://github.com/epinowcast/epidist/pull/390/commits/e2e68d245e2caf8e2b7f6877d7238898b1fc5ba8)

[the remark this answers is not in the document]
Comment (athowes) [reply NAMESPACE]: Issue for new method is here https://github.com/epinowcast/epidist/issues/414

File: R/direct_model.R [hunk 5]
```diff
@@ -35,23 +24,14 @@ as_direct_model.data.frame <- function(data) {
...
-#' with the correct columns.
-#'
-#' @param data A `data.frame` containing line list data
-#' @param ... ...
-#' @method epidist_validate epidist_direct_model
+#' @method epidist_validate_model epidist_direct_model
 #' @family direct_model
 #' @export
-epidist_validate.epidist_direct_model <- function(data, ...) {
+epidist_validate_model.epidist_direct_model <- function(data, ...) {
```
Comment (seabbs) [inline R/direct_model.R:34]: you can just stick to having one validate function as the class of input can be used i.e direct_model vs linelist

File: R/latent_individual.R [hunk 6]
```diff
@@ -7,89 +7,42 @@ as_latent_individual <- function(data) {
...
+  col_names <- c(
+    "ptime_lwr", "ptime_upr", "stime_lwr", "stime_upr", "obs_time",
+    "relative_obs_time", "pwindow", "woverlap", "swindow", "delay", ".row_id"
   )
-  if (nrow(data) > 1) {
-    assert_factor(data$row_id)
-  }
+  assert_names(names(data), must.include = col_names)
   assert_numeric(data$relative_obs_time, lower = 0)
+  # pwindow as f(p) and swindow as f(s) checks here?
```
Comment (seabbs) [inline R/latent_individual.R:45]: remove comment question?

**Review by seabbs — APPROVED** [review]
This looks good. 

I think I would have just a single validate function and dispatch into it from models and data. In the future we should also update to use modification validation vs entry to use validation. We should also add a new_epidist_linelist method in the future as well.

**Pull request merged**